Pyodide 0.27 Release Notes (#48)
pyodide/pyodide-blog: content/posts/0.27-release.md

# Welcome Agriya Khetarpal to the Pyodide team

Agriya Khetarpal has joined Pyodide as the new maintainer.
Agriya has been active in the Scientific Python area and is a contributor of Numpy, Scipy, and scikit-learn.
He has already contributed significantly to strengthening Pyodide support in various Scientific Python packages.

# Pyodide 0.27 Release

## Build System Improvements

### Decoupling pyodide-build from Pyodide runtime

pyodide-build is a tool that builds Python packages for the WASM environment to run in Pyodide.
Previously, the version of pyodide-build was strongly tied to the version of Pyodide,
meaning that if you wanted to build a package against a specific version of Pyodide, you had to use a corresponding version of pyodide-build.
The problem with this approach was that even if we improved the build system, you would have to wait a few months after Pyodide was released to use it.

In this release, we have separated pyodide-build from the Pyodide runtime.
This allows us to develop and release pyodide-build independently of Pyodide more quickly,
and you can apply these improvements to your packages sooner.

> pyodide-build is now developed in [pyodide/pyodide-build](https://github.com/pyodide/pyodide-build),
> and you can install it with the command `pip install pyodide-build`.

Reviewer: Could we add a compatibility note here? Please feel free to suggest something better.

````suggestion
pyodide-build is now developed in [pyodide/pyodide-build](https://github.com/pyodide/pyodide-build),
and you can install it with the command `pip install pyodide-build`.

Recent versions of `pyodide-build` work with Pyodide 0.26 and higher. This means you can use, say,
`pyodide-build` version 0.29.2, to set up an older Pyodide 0.26.4 cross-build environment:

```bash
pip install pyodide-build==0.29.2
pyodide xbuildenv install 0.26.4
pyodide build <...>
```

In case we decide to drop compatibility for a particular version of Pyodide, we shall mark it as such in
the [cross-build environments metadata file](https://github.com/pyodide/pyodide/blob/main/pyodide-cross-build-environments.json).
````

Author: Sounds good. Thanks!

### A potential future performance improvement

In the case of lookup on a `JsProxy`, there are other bytecode patterns where we
could optimize away temporaries. For instance, if `a` is a `JsProxy` for the
JavaScript object `{ x: { y: 2 } }` then `a.x.y` creates a temporary `JsProxy`
pointing to `{ y: 2 }` and then looks up `y` on the temporary and returns `2`.
The bytecode looks like:
```
LOAD a
LOAD_ATTR x
LOAD_ATTR y
```
If we know that we were going to use the result of the first `LOAD_ATTR x`
immediately, we could avoid creating the temporary. Unfortunately, unlike the
method call example, native Python code doesn't create a temporary here so
optimizing this would involve more invasive patches to the interpreter than
removing the temporary bound methods.

## Work in Progress / Roadmap

TODO

## Acknowledgements

Thanks to Agriya Khetarpal, Loïc Estève, and Ralf Gommers for their work helping ensure scientific Python packages are well supported in Pyodide.
Thanks to Joe Marshall for his huge contribution on PyArrow Support.

Additionally, we always appreciate the support and assistance from the Emscripten team.

> The following people commited to Pyodide in this release:
> 
> Agriya Khetarpal, airen1986, Andrei V. Plamada, Andrew Moon, Bart Broere, Carlo 
> Piovesan, Castedo Ellerman, Chris Pyles, Christian Clauss, chrysn, Deepak 
> Cherian, Eli Lamb, Em Zhan, Eric Brown, Gyeongjae Choi, Hanno Rein, Henry 
> Schreiner, Hood Chatham, Ian Thomas, James J Balamuta, JHM Darbyshire, Jiefu7, 
> Joel Ostblom, Joe Marshall, josephrocca, Juniper Tyree, Kellen Malek, Kyle 
> Barron, Loïc Estève, Luiz Irber, Maarten Breddels, Marco Edward Gorelli, 
> Marianne Corvellec, M Bussonnier, Muspi Merol, Myles Scolnick, Nick Altmann, 
> Olivier Grisel, Oscar Benjamin, Péter Gyarmati, Phillip Cloud, pyodide-pr-bot, 
> Riya Sinha, swnf, Szabolcs Dombi, Victor Blomqvist, Yan Wong, YISH, Zsolt

Reviewer: Here is an updated list from earlier today:

```suggestion
The following people committed to Pyodide in this release:

Agriya Khetarpal, Andrei V. Plamada, Andrew Moon, Bart Broere, Carlo
Piovesan, Castedo Ellerman, Chris Pyles, Christian Clauss, Deepak Cherian,
Eli Lamb, Em Zhan, Eric Brown, Gyeongjae Choi, Hanno Rein, Henry Schreiner,
Hood Chatham, Ian Thomas, JHM Darbyshire, James J Balamuta,
Jiefu7, Joe Marshall, Joel Ostblom, Juniper Tyree, Kellen Malek, Kyle Barron,
Loïc Estève, Luiz Irber, M Bussonnier, Maarten Breddels, Marco Edward Gorelli,
Marianne Corvellec, Muspi Merol, Myles Scolnick, Nick Altmann, Olivier Grisel,
Oscar Benjamin, Péter Gyarmati, Phillip Cloud, Riya Sinha, Szabolcs Dombi,
Victor Blomqvist, YISH, Yan Wong, Zsolt Dollenstein, airen1986, chrysn,
josephrocca, swnf
```

I used this command to calculate and sort it:

```bash
git log --pretty=format:"%an" 0.26.0..HEAD \
| sort -u | grep -v "\[bot\]\|pyodide-pr-bot" | \
paste -sd "," - | sed 's/,/, /g'
```

example, you can use `pyodide-build` version 0.29.2 to build packages for use
with Pyodide 0.26.4:

```bash
pip install pyodide-build==0.29.2
pyodide xbuildenv install 0.26.4
pyodide build <...>
```

### Wheels for new packages

This release includes about twenty new packages, most notably popular data
science packages PyArrow, Polars, and DuckDB.

Previously, we have built all packages in our CI from within the repository. The
increasing number of packages with long build times have been a strain on our CI
resources. For instance, PyArrow takes tens of minutes to build. Now, several of
these packages are built in a separate repository and managed by the package
maintainers. In Pyodide 0.28.0,
[we are planning to unvendor the packages from the Pyodide runtime entirely](https://github.com/pyodide/pyodide/issues/4918).
This will pave the way to supporting many more packages.

### Update to NumPy 2.0

We have updated NumPy to version 2.0.2. This is a major update that required
updates to many downstream packages in the scientific Python ecosystem. Please
see [Ecosystem compatibility with numpy 2.0](https://github.com/numpy/numpy/issues/26191)

> for information.

Reviewer: Oops, thanks!

> We have updated NumPy to version 2.0.2. This is a major update that required
> updates to many downstream packages in the scientific Python ecosystem. Please
> see [Ecosystem compatibility with numpy 2.0](https://github.com/numpy/numpy/issues/26191)
> for information about downstream compatibility with NumPy 2.0

Reviewer: ```suggestion
We have updated NumPy to version 2.0.2. This is a major update that required
updates to many downstream packages in the scientific Python ecosystem. Please
see [Ecosystem compatibility with numpy 2.0](https://github.com/numpy/numpy/issues/26191)
for information about downstream compatibility with NumPy 2.0.
```

Also, the word "compatibility" comes up twice, could we reword this sentence?

Reviewer 2: Yeah just say "for more information"

Reviewer: Makes sense, addressed in a4b6d372cd53621300a2bbd91095c0fad1a3d4ea.

JavaScript object `{ x: { y: 2 } }` then `a.x.y` creates a temporary `JsProxy`
pointing to `{ y: 2 }` and then looks up `y` on the temporary and returns `2`.
The bytecode looks like:
```
LOAD a
LOAD_ATTR x
LOAD_ATTR y
```
If we know that we were going to use the result of the first `LOAD_ATTR x`
immediately, we could avoid creating the temporary. Unfortunately, unlike the
method call example, native Python code doesn't create a temporary here so
optimizing this would involve more invasive patches to the interpreter than
removing the temporary bound methods.

## Work in Progress / Roadmap

For Pyodide 0.28, we hope to unvendor the package recipes from the main Pyodide
repository. The package index will gain a separate release process from the
Pyodide runtime itself. This should have several significant benefits:

* It will reduce CI usage due to rebuilding packages less often.
* People who add a package will not have to wait as long until their package is
  included in our index.

We are working on adding Emscripten back as a tier 3 target for Python. As a
part of this work, we have recently upstreamed a large number of fixes to the

> Emscripten file system. This will make a lot of obscure h

Reviewer: @hoodmane, was this sentence missed when you were writing it, by chance?

Reviewer 2: I just pushed my local state as-is because I was not actively working on it and wanted to avoid conflicts.

stability of Pyodide.

# Welcome Agriya Khetarpal to the Pyodide team

Agriya Khetarpal has joined Pyodide as the new maintainer. Agriya has been
active in the Scientific Python area and is a contributor of NumPy, SciPy, and
scikit-learn. He has already contributed significantly to strengthening Pyodide
support in various Scientific Python packages.

# Pyodide 0.27 Release

## Build System Improvements

### Decoupling `pyodide-build` from Pyodide runtime

`pyodide-build` is a tool that builds Python packages for the WASM environment to
run in Pyodide. Previously, the version of `pyodide-build` was strongly tied to
the version of Pyodide, meaning that if you wanted to build a package against a
specific version of Pyodide, you had to use a corresponding version of
`pyodide-build`. The problem with this approach was that even if we improved the
build system, you would have to wait a few months after Pyodide was released to
use it.

In this release, we have separated `pyodide-build` from the Pyodide runtime.
This allows us to develop and release `pyodide-build` independently of Pyodide
more quickly, and you can apply these improvements to your packages sooner.

> pyodide-build is now developed in [pyodide/pyodide-build](https://github.com/pyodide/pyodide-build),

Reviewer: ```suggestion
`pyodide-build` is now developed in [pyodide/pyodide-build](https://github.com/pyodide/pyodide-build),
```

and you can install it with the command `pip install pyodide-build`.

Recent versions of `pyodide-build` work with Pyodide 0.26 and higher. This means
you can use, say, `pyodide-build` version 0.29.2, to set up an older Pyodide
0.26.4 cross-build environment:

```bash
pip install pyodide-build==0.29.2
pyodide xbuildenv install 0.26.4
pyodide build <...>
```

> In case we decide to drop compatibility for a particular version of Pyodide, we
> shall mark it as such in the 
> [cross-build environments metadata file](https://github.com/pyodide/pyodide/blob/main/pyodide-cross-build-environments.json).

Reviewer: ```suggestion
In case we decide to drop compatibility for a particular version of Pyodide, we
shall mark it as such in the 
[cross-build environments metadata file](https://github.com/pyodide/pyodide/blob/main/pyodide-cross-build-environments.json)
and announce it with the release notes.
```

> [cross-build environments metadata file](https://github.com/pyodide/pyodide/blob/main/pyodide-cross-build-environments.json).

Reviewer 2: Can we make this sentence more specific? It seems to me that each Pyodide version requires a minimum Pyodide build version, but I don't see a field to specify a maximum Pyodide build version. @agriyakhetarpal do we have a field for that and it's just absent? What will we do if we drop compatibility for a version of Pyodide?

Reviewer: Yes, we don't have one for a maximum version yet, because it is perhaps implied that there isn't an "unreleased" Pyodide version that a released version of `pyodide-build` wouldn't support out of the box (we'll catch it in CI). I think we should add one, even if we won't use it actively.

https://github.com/pyodide/pyodide-build/issues/77

Reviewer 2: I think we should delete this sentence because it's not clear to a user what it means. Why do they care that it's marked in the cross-build environments metadata file? Presumably what they care about is that it will be detected if they have a version mismatch. But isn't that assumed?

Reviewer: Yes, makes sense to me.

### Wheels for new packages

> This release includes about a dozen new packages, most notably PyArrow, Polars,
> and DuckDB, which people have long requested. What's unique about these new

Reviewer: ```suggestion
This release includes about a dozen new packages, most notably PyArrow, Polars,
and DuckDB, which users have long requested. What's unique about these new
```

`type(a).f(a)`. The `LOAD_ATTR` opcode has a special argument that indicates
that the next opcode is going to be `CALL` and in that case calls a method named
`_PyObject_GetMethod` instead of the typical `PyObject_GetAttr` to perform the
attribute lookup. We patched `_PyObject_GetMethod` to have special handling for
`JsProxy` objects so that we could optimize away the temporary for JS objects too.

### A potential future performance improvement

In the case of lookup on a `JsProxy`, there are other bytecode patterns where we
could optimize away temporaries. For instance, if `a` is a `JsProxy` for the
JavaScript object `{ x: { y: 2 } }` then `a.x.y` creates a temporary `JsProxy`
pointing to `{ y: 2 }` and then looks up `y` on the temporary and returns `2`.
The bytecode looks like:
```
LOAD a
LOAD_ATTR x
LOAD_ATTR y
```
If we know that we were going to use the result of the first `LOAD_ATTR x`
immediately, we could avoid creating the temporary. Unfortunately, unlike the
method call example, native Python code doesn't create a temporary here so
optimizing this would involve more invasive patches to the interpreter than
removing the temporary bound methods.

## What's next?

> Our next release should use Python 3.13. In addition, we are planning to make
> the following changes:

Reviewer: ```suggestion
Our next release [should use Python 3.13](https://github.com/pyodide/pyodide/pull/5205). In addition, we are planning to make
the following changes:
```

Reviewer 2: I think we can leave this link out.

### Package unvendoring

We will unvendor the package recipes from the main Pyodide repository. The
package index will gain a separate release process from the Pyodide runtime
itself. This should have several significant benefits:

* It will reduce CI usage due to rebuilding packages less often.
* People who add a package will not have to wait as long until their package is
  included in our index.

### Upstream work in Emscripten and Python

We are working on restoring Emscripten to a tier 3 supported target for Python.
As a part of this work, we are upstreaming a large number of fixes to the
Emscripten file system. For example, it now works to seek on `/dev/null` (it
does nothing), symlink support is much better, it will work to `stat` a file
descriptor which doesn't point to a named file, and many file system system

> calls have more posix-compliant error handling. Most of these changes have not
> made their way into Pyodide yet because we're still using Emscripten 3.1.68 for
> ABI compatibility.

Reviewer: ```suggestion
calls have more posix-compliant error handling. Most of these changes have not
made their way into Pyodide yet because we're still using Emscripten 3.1.58 for
ABI compatibility.
```

(probably a typo?)

using Wasm Exception Handling for C++ exceptions, Rust panics, and
setjmp/longjmp. This will lead to faster, smaller code, and fewer bugs. It also
eliminates a lot of complexity in the stack switching support code -- it is
impossible to stack switch through JavaScript frames and by default every C++
try block introduces a JavaScript frame. This requires upstream work on the Rust
compiler and switching to a custom build of the Rust standard library.

## Acknowledgements

Thanks to Agriya Khetarpal, Loïc Estève, and Ralf Gommers for their work helping
ensure that packages in the Scientific Python ecosystem are well supported in
Pyodide.

Thanks to [Joe Marshall](https://people.cs.nott.ac.uk/pszjm2/about/) and
[George Stagg](https://gws.phd/) for their contributions towards
[PyArrow](https://github.com/apache/arrow/pull/37822) and
[Polars](https://github.com/pola-rs/polars/pull/20383) support respectively, and
to DuckDB maintainers for
[getting DuckDB to work in Pyodide](https://duckdb.org/2024/10/02/pyodide.html).

Additionally, we appreciate the continued support from the Emscripten team.

The following people committed to Pyodide in this release:

> Agriya Khetarpal, Andrei V. Plamada, Andrew Moon, Bart Broere, Carlo
> Piovesan, Castedo Ellerman, Chris Pyles, Christian Clauss, Deepak Cherian,
> Eli Lamb, Em Zhan, Eric Brown, Gyeongjae Choi, Hanno Rein, Henry Schreiner,
> Hood Chatham, Ian Thomas, JHM Darbyshire, James J Balamuta, James Lamb,
> Jiefu7, Joe Marshall, Joel Ostblom, Juniper Tyree, Kellen Malek,
> Kyle Barron,Loïc Estève, Luiz Irber, M Bussonnier, Maarten Breddels,
> Marco Edward Gorelli, Marianne Corvellec, Muspi Merol, Myles Scolnick,
> Nick Altmann, Olivier Grisel, Oscar Benjamin, Péter Gyarmati, Phillip Cloud,
> Riya Sinha, Szabolcs Dombi, Victor Blomqvist, YISH, Yan Wong,
> Zsolt Dollenstein, airen1986, chrysn, josephrocca, swnf

Reviewer: ```suggestion
Agriya Khetarpal, Andrei V. Plamada, Andrew Moon, Bart Broere, Carlo
Piovesan, Castedo Ellerman, Chris Pyles, Christian Clauss, Deepak Cherian,
Eli Lamb, Em Zhan, Eric Brown, Gyeongjae Choi, Hanno Rein, Henry Schreiner,
Hood Chatham, Ian Thomas, JHM Darbyshire, James J Balamuta, James Lamb,
Jiefu7, Joe Marshall, Joel Ostblom, Juniper Tyree, Kellen Malek,
Kyle Barron, Loïc Estève, Luiz Irber, M Bussonnier, Maarten Breddels,
Marco Edward Gorelli, Marianne Corvellec, Muspi Merol, Myles Scolnick,
Nick Altmann, Olivier Grisel, Oscar Benjamin, Péter Gyarmati, Phillip Cloud,
Riya Sinha, Szabolcs Dombi, Victor Blomqvist, YISH, Yan Wong,
Zsolt Dollenstein, airen1986, chrysn, josephrocca, swnf
```

---
title: "Pyodide 0.27 Release"
date: 2024-09-27T11:00:55-04:00
draft: true
tags: ["announcement"]
# author: "Me"
author: ["Gyeongjae Choi"] # multiple authors
showToc: true
TocOpen: false
draft: false
hidemeta: false
comments: false
# canonicalURL: "https://canonical.url/to/page"
disableHLJS: true # to disable highlightjs
disableShare: false
hideSummary: false
searchHidden: true
ShowReadingTime: true
ShowBreadCrumbs: true
ShowPostNavLinks: true
cover:
    # image: "<image path/url>" # image path/url
    # alt: "<alt text>" # alt text
    # caption: "<text>" # display caption under cover
    relative: false # when using page bundles set this to true
    hidden: true # only hide on current single page
---

Pyodide v0.27.0 is out. This release focuses on improving the long-term stability of Pyodide.

# Welcome Agriya Khetarpal to the Pyodide team

Agriya Khetarpal has joined Pyodide as the new maintainer.
Agriya has been active in the Scientific Python area and is a contributor of Numpy, Scipy, and scikit-learn.
He has already contributed significantly to strengthening Pyodide support in various Scientific Python packages.

# Pyodide 0.27 Release

## Build System Improvements

> ### Decoupling pyodide-build from Pyodide runtime

Reviewer: Maybe we could add the relevant links (PyPI, GitHub, and more) for `pyodide-build` so that users can access them from here?

Author: Sounds good.

meaning that if you wanted to build a package against a specific version of Pyodide, you had to use a corresponding version of pyodide-build.
The problem with this approach was that even if we improved the build system, you would have to wait a few months after Pyodide was released to use it.

In this release, we have separated pyodide-build from the Pyodide runtime.
This allows us to develop and release pyodide-build independently of Pyodide more quickly,
and you can apply these improvements to your packages sooner.

### New Packages Support

This release includes about a dozen new packages, most notably PyArrow and DuckDB, which people have long requested.
What's unique about these new packages is that they are not being built from the Pyodide repository,
but are being built out-of-tree by the package maintainers using pyodide-build.

Previously, Pyodide has built all packages directly from within the repository.
This had been a significant limitation on long-term scalability,
especially for large libraries like PyArrow, which take tens of minutes or more to build.
Now, these packages are built in a separate repository and managed by the package maintainers.

> We expect this change will improve package stability and pave the way for more packages to be added to Pyodide in the long run.

Reviewer: It might be good to include a link to https://github.com/pyodide/pyodide/issues/4918 here at the end, since it is an RFC issue and we would appreciate thoughts from other people.

## Performance improvements

JsProxy

## Work in Progress / Roadmap

TODO

## Acknowledgements

> Thanks to Agriya Khetarpal, Loïc Estève, and Ralph Gommers for their work helping ensure scientific Python packages are well supported in Pyodide.

Reviewer: ```suggestion
Thanks to Agriya Khetarpal, Loïc Estève, and Ralf Gommers for their work helping ensure scientific Python packages are well supported in Pyodide.
```

I think this was incorrect in the notes for the previous release as well (so maybe we can fix that here)?

Author: Ohhh nice catch, thanks!

and you can apply these improvements to your packages sooner.

pyodide-build is now developed in [pyodide/pyodide-build](https://github.com/pyodide/pyodide-build),
and you can install it with the command `pip install pyodide-build`.

### New Packages Support

This release includes about a dozen new packages, most notably PyArrow and DuckDB, which people have long requested.
What's unique about these new packages is that they are not being built from the Pyodide repository,
but are being built out-of-tree by the package maintainers using pyodide-build.

Previously, Pyodide has built all packages directly from within the repository.
This had been a significant limitation on long-term scalability,
especially for large libraries like PyArrow, which take tens of minutes or more to build.
Now, these packages are built in a separate repository and managed by the package maintainers.
We expect this change will improve package stability and [pave the way](https://github.com/pyodide/pyodide/issues/4918) for more packages to be added to Pyodide in the long run.

## Performance improvements

JsProxy

## Work in Progress / Roadmap

> TODO

Author: @hoodmane Do you want to add some content on these sections?

Reviewer 2: Yeah. I have been very busy for a few weeks, but I'm mostly back to my normal schedule now. I think I'll have time to work on this on Tuesday.
